// ngx-bootstrap/utils/esm2022/theme-provider.mjs
import { window } from './facade/browser';
export var BsVerions;
(function (BsVerions) {
    BsVerions["isBs4"] = "bs4";
    BsVerions["isBs5"] = "bs5";
})(BsVerions || (BsVerions = {}));
let guessedVersion;
function _guessBsVersion() {
    const spanEl = window.document.createElement('span');
    spanEl.innerText = 'testing bs version';
    spanEl.classList.add('d-none');
    spanEl.classList.add('pl-1');
    window.document.head.appendChild(spanEl);
    const checkPadding = window.getComputedStyle(spanEl).paddingLeft;
    if (checkPadding && parseFloat(checkPadding)) {
        window.document.head.removeChild(spanEl);
        return 'bs4';
    }
    window.document.head.removeChild(spanEl);
    return 'bs5';
}
export function setTheme(theme) {
    guessedVersion = theme;
}
export function isBs4() {
    if (guessedVersion)
        return guessedVersion === 'bs4';
    guessedVersion = _guessBsVersion();
    return guessedVersion === 'bs4';
}
export function isBs5() {
    if (guessedVersion)
        return guessedVersion === 'bs5';
    guessedVersion = _guessBsVersion();
    return guessedVersion === 'bs5';
}
export function getBsVer() {
    return {
        isBs4: isBs4(),
        isBs5: isBs5()
    };
}
export function currentBsVersion() {
    const bsVer = getBsVer();
    const resVersion = Object.keys(bsVer).find(key => bsVer[key]);
    return BsVerions[resVersion];
}
